integration_test.go

```diff
@@ -67,36 +62,90 @@ func runsOnEmulator() bool {
+	_, err = instanceAdmin.GetInstance(ctx, &instancepb.GetInstanceRequest{
+		Name: fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId),
+	})
+	if err == nil {
+		return func() {}, nil
+	}
+	if spanner.ErrCode(err) != codes.NotFound {
+		return nil, err
+	}
+
+	// Instance does not exist. Create a temporary instance for this test run.
+	// The instance will be deleted after the test run.
 	op, err := instanceAdmin.CreateInstance(ctx, &instancepb.CreateInstanceRequest{
 		Parent:     fmt.Sprintf("projects/%s", projectId),
 		InstanceId: instanceId,
 		Instance: &instancepb.Instance{
-			Config:      fmt.Sprintf("projects/%s/instanceConfigs/%s", projectId, "emulator-config"),
+			Config:      fmt.Sprintf("projects/%s/instanceConfigs/%s", projectId, config),
 			DisplayName: instanceId,
 			NodeCount:   1,
+			Labels: map[string]string{
+				"gosqltestinstance": "true",
+				"createdat":         fmt.Sprintf("t%d", time.Now().Unix()),
+			},
 		},
 	})
 	if err != nil {
-		// Check if the instance has already been created by another (parallel) test.
-		code := spanner.ErrCode(err)
-		if code != codes.AlreadyExists {
-			return fmt.Errorf("could not create instance %s: %v", fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId), err)
-		}
+		return nil, fmt.Errorf("could not create instance %s: %v", fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId), err)
 	} else {
 		// Wait for the instance creation to finish.
 		_, err := op.Wait(ctx)
 		if err != nil {
-			return fmt.Errorf("waiting for instance creation to finish failed: %v", err)
+			return nil, fmt.Errorf("waiting for instance creation to finish failed: %v", err)
 		}
 	}
-	return nil
+	// Delete the instance after all tests have finished.
+	// Also delete any stale test instances that might still be around on the project.
+	return func() {
+		instanceAdmin, err := instance.NewInstanceAdminClient(ctx)
+		if err != nil {
+			return
+		}
+		// Delete this test instance.
+		instanceAdmin.DeleteInstance(ctx, &instancepb.DeleteInstanceRequest{
+			Name: fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId),
+		})
+		// Also delete any other stale test instance.
+		instances := instanceAdmin.ListInstances(ctx, &instancepb.ListInstancesRequest{
+			Parent: fmt.Sprintf("projects/%s", projectId),
+			Filter: "label.gosqltestinstance:*",
+		})
+		for {
+			instance, err := instances.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				log.Printf("failed to fetch instances during cleanup: %v", err)
+				break
+			}
+			if createdAtString, ok := instance.Labels["createdat"]; ok {
+				// Strip the leading 't' from the value.
+				seconds, err := strconv.ParseInt(createdAtString[1:], 10, 64)
+				if err != nil {
+					log.Printf("failed to parse created time from string %q of instance %s: %v", createdAtString, instance.Name, err)
+				} else {
+					diff := time.Duration(time.Now().Unix()-seconds) * time.Second
+					if diff > time.Hour*24 {
```

Comment:
WDYT about using a shorter time difference to prevent too many instances being deleted at once and hitting the admin requests per minute limit? Since the limit is at the project level it could cause failures for other test runs. In Python, we use one hour but that should probably be bumped to at least two hours.

Reply:
I've reduced it to two hours, as I think you're right that having them for too long does not add any value anyways.

I don't really worry about the administrative requests limit here, as a temporary instance will normally be deleted after the tests have been run, and one test run will create at most one instance. The administrative request limit will probably be more of a problem because we are executing the integration tests (partly) in parallel, and each test creates its own temporary database. If we start hitting that limit, we probably need to add a retryer for that specific error.